### ISSUE TITLE
Confusing mesh result on nerd_ehead

### ISSUE THREAD

**Issue by YuhsiHu** [NVlabs/nvdiffrec#11]
Thank you for the great work! 
I followed the one time setup on ubuntu 18.04. Python=3.9, CUDA=11.3, PyTorch=1.10.
Then I had a trial on nerd_ehead by running:
`python train.py --config configs/nerd_ehead.json`
In **dmtet_mesh** folder, the mesh looks good :
![image](https://user-images.githubusercontent.com/34529345/161021375-fa12924c-7376-40a4-92fb-eacbe3716c75.png)
But in **mesh** folder, the **mesh.obj** looks confusing:
![image](https://user-images.githubusercontent.com/34529345/161021600-c1575283-a755-46c7-abc0-ad90f1c2fe97.png)
I did not change any code in this repo so I did not know why this happened. Could you please help me solve this problem? Thank you for your time!

**Comment by jmunkberg** [comment]
Hello, 

I just tried it on Ubuntu 20.04.3 LTS, and got this in the mesh folder
![image](https://user-images.githubusercontent.com/68556069/161041709-59211470-a57a-4553-a201-3d64223a5c96.png)
That particular scene varies a bit between runs, but we haven't seen a collapse like the one above. 

Did you run the nerd_ehead.json config unmodified or with reduced batch size? Are the other configs working on your machine?

**Comment by JHnvidia** [comment]
I also haven't seen it, so it could just be an divergence in early training. You could try and see what the `img_mesh_pass*.png` files look like. If the problem persists you can try and add a line `"lock_pos" : true` to the `.json` config file as a workaround. This will disable position training/tuning during the second (mesh) pass, and should ensure you get the same geometry from both passes.

**Comment by YuhsiHu** [comment]
Here is the  **img_mesh_pass_000000.jpg**:
![img_mesh_pass_000000](https://user-images.githubusercontent.com/34529345/161180148-fce9006d-6add-46b7-8ee9-482f8bec382a.png)
And the **img_mesh_pass_000001.jpg** looks like:
![img_mesh_pass_000001](https://user-images.githubusercontent.com/34529345/161180231-f2f58a6a-df70-4c92-9a95-d751572c9a13.png)
I will try add this line and update the progress. Thank you!

**Comment by YuhsiHu** [comment]
Add `"lock_pos":true` to the `json` file did avoid that. However, the PSNR became lower.
```
MSE,      PSNR
0.00195455, 27.823
Base mesh has 7708 triangles and 3862 vertices.
```
```
MSE,      PSNR 
0.00389249, 24.451
Writing mesh:  out/nerd_ehead/mesh/mesh.obj
    writing 3862 vertices
    writing 6643 texcoords
    writing 3862 normals
    writing 7708 faces
Writing material:  out/nerd_ehead/mesh/mesh.mtl
Done exporting mesh
```
Other results are good(e.g. chair). Why did this happen?

**Comment by jmunkberg** [comment]
Just to confirm, you are indeed running with batch size 8 in the config? Smaller batch size means more noise in the gradients, and can be harder to train. 

Another thing you could try is to instead of using `"lock_pos" : true` , you can lower the learning rate in the second pass. So replace the line 
`"learning_rate": [0.03, 0.03],` with, say, `"learning_rate": [0.03, 0.01],` or even `"learning_rate": [0.03, 0.003],`

That said, the default config should produce better results. I checked a recent run with the default config unmodified, and these are the first two frames from the img_mesh_pass from that particular run. 
![img_mesh_pass_000000](https://user-images.githubusercontent.com/68556069/161196616-ce17a4cb-8efb-4759-a93d-79d314f70eea.png)
![img_mesh_pass_000001](https://user-images.githubusercontent.com/68556069/161196633-c239d139-0899-404c-8fbf-96b8029d952d.png)

**Comment by YuhsiHu** [comment]
I did not modify batch size or any other parameter. The first two frames from **img_mesh_pass** are as follows:
![img_mesh_pass_000000](https://user-images.githubusercontent.com/34529345/161206301-6786a3c4-1a51-4cdb-b090-8e6acf478900.png)
![img_mesh_pass_000001](https://user-images.githubusercontent.com/34529345/161206331-acebaae4-8880-4121-94f5-775722d1191a.png)

**Comment by jmunkberg** [comment]
This looks much better! Are you saying that one particular training diverged and that it is now working on your end? We are seeing some variation from run to run (the SDF is randomly initialized etc.) but on the two runs we started this morning on two different machines (using the released code unmodified), both looked as expected. Here is img_mesh_pass_000050.png

![img_mesh_pass_000050](https://user-images.githubusercontent.com/68556069/161207188-cc1a5aa0-66f0-4d7b-83f5-743ba0010724.png)

**Comment by JHnvidia** [comment]
I've also re-ran the `nerd_ehead.json` config and it seems to behave ok on my end (first two **img_mesh_pass**).

![img_mesh_pass_000000](https://user-images.githubusercontent.com/40264125/161206087-2a5519c7-53d2-4e22-ba85-961f18dfdc79.png)
![img_mesh_pass_000001](https://user-images.githubusercontent.com/40264125/161206091-58a3f666-ff47-409a-a2b4-0325cbd25619.png)

Please try a fresh `git clone`, re-run the unmodified config and see if the problem persists. Since the training is stochastic there's a small chance that it's just some random fluke.

**Comment by YuhsiHu** [comment]
Thank you for your reply. I tried git clone again but there was no change on my machine. The result is the same if I did not add `"lock_pos" : true` to `json` file.  Maybe I should set the learning rate lower?

**Comment by jmunkberg** [comment]
Yes, as discussed above, you can try reducing the learning rate, but the default config should generate better results than what your report. We tried reproducing your issue on two different machines, using the code and config from the repo unmodified and we get much better results, as posted above. Some variations between runs is expected, but I would expect results in line with what we posted above.
